subsys: usb: host: Add USB Host CDC ECM Class Support for Ethernet

santhosh-c-c opened the pull request:
USB Host: Add CDC ECM class support

This PR introduces the USB CDC ECM (Ethernet Control Model) class support 
to the Zephyr USB host subsystem, enabling Ethernet functionality for 
USB hosts.

Changes since draft:
- Rebased on merged dependencies #94590
- Subsys: Optimized High-Speed device initialization by defaulting 
  bMaxPacketSize0 to 64 per USB 2.0 spec (85df60a)

Dependency:
- #106802 (usb: host: fix HS device enumeration and interface probing) carries equivalent and additional USBH core fixes needed for HS device interop. The bMaxPacketSize0 commit in this PR overlaps with the first commit of 106802 and will be dropped from this PR once 106802 merges.

Signed-off-by: Santhosh Charles <santhosh@linumiz.com>

santhosh-c-c committed:
usb: host: class: add CDC ECM class support for USBH

Add support for the USB CDC ECM (Ethernet Control Model) class to the
USB host subsystem. This implementation enables Ethernet functionality
for USB host.

Signed-off-by: Santhosh Charles <santhosh@linumiz.com>

santhosh-c-c commented:
> > @MarkWangChinese is this related to your ongoing effort for implementing ECM?
> 
> Thanks @josuah
> 
> Hi @santhosh-c-c As the comments [#95661 (comment)](https://github.com/zephyrproject-rtos/zephyr/issues/95661#issuecomment-3490731892), our side (NXP) is trying to enable the USB Host ECM too, our plan is creating the formal pr before the end of December, we will do basic function test too. Could you tell me your plan? let's align to meet both our requirements. Thanks.

Hello @MarkWangChinese

Our code has been tested on the hardware (MIMXRT1064-EVK), and we are currently waiting for the dependency to be merged first, namely this PR: https://github.com/zephyrproject-rtos/zephyr/pull/94590. Once that merge is complete, we will proceed with merging our ECM host class support.

An internal review is ongoing and we are looking for any features that should be added. This internal review should not take more than a week. If you would like, we can discuss any additional features that need to be added so we can align with your requirements.

carlescufi commented:
@santhosh-c-c dependencies should now be merged. Please take out of draft and rebase on top of `main`.

D-Veda committed:
usb: host: ch9: Add string descriptor related request helper function

Add helpers to get USB device string descriptors, obtain a list of the
supported LANGIDs, and convert UTF16LE encoded string descriptors to
ASCII.

Co-authored-by: Johann Fischer <johann.fischer@nordicsemi.no>
Signed-off-by: Dv Alan <zhangyang.shen@nxp.com>

santhosh-c-c commented:
@josuah @carlescufi @MarkWangChinese 
Ping for initial review, Thanks!

santhosh-c-c commented:
Hello @jfischer-no @josuah @MarkWangChinese Pinging this PR for review as discussed in the other thread https://github.com/zephyrproject-rtos/zephyr/issues/95661#issuecomment-4273996267. I'm standing by to address any comments or requested changes to help get this core support merged.

josuah reviewed (commented):
It seems like this PR is smaller, simpler which makes it easier to integrate.

With https://github.com/zephyrproject-rtos/zephyr/pull/100289 already going on, this means that some deduplication needs to happen, and using the current PR as a basis for https://github.com/zephyrproject-rtos/zephyr/pull/100289 is not feasible given the difference of structure.

Do you think it would be possible to refactor this PR to make it closer to #100289 while also applying your subjective improvements?

For instance, every time something is imported with changes applied, also propose the changes in #100289 as a review.

josuah commented:
It is possible to merge some commits from this PR outside of CDC ECM context.

It is possible to integrate some commits from https://github.com/zephyrproject-rtos/zephyr/pull/100289 into this current PR.

This directly contributes to progress on CDC ECM host support.

Thank you for the efforts and time on this.

santhosh-c-c commented:
> It seems like this PR is smaller, simpler which makes it easier to integrate.
> 
> With #100289 already going on, this means that some deduplication needs to happen, and using the current PR as a basis for #100289 is not feasible given the difference of structure.
> 
> Do you think it would be possible to refactor this PR to make it closer to #100289 while also applying your subjective improvements?
> 
> For instance, every time something is imported with changes applied, also propose the changes in #100289 as a review.

Thanks @josuah for the input. A few clarifications

1. On refactoring this PR closer to #100289: the structural difference is intentional. This PR keeps the implementation minimal and focused on the core ECM data path. #100289 layers significantly more on top (multicast filter, packet filter bitmap, hw stats, link speed, async xfer queue with cleanup list). Refactoring this PR to mirror that structure would essentially turn it into a duplicate of #100289 and defeat the purpose of having a small, reviewable base.

2. On commits outside the CDC ECM context: agreed. The `usb: host: set default HS bMaxPacketSize0 to 64` commit from this PR is a generic USBH core fix unrelated to ECM and is already being addressed independently in #106802 (along with two other HS enumeration/probing fixes from the same testing effort). I will drop that commit from this PR once #106802 lands.

3. On integrating commits from #100289 into this PR: this is the part of your suggestion I want to act on directly. The two non-driver commits in #100289 are useful regardless of which PR lands the ECM driver,  the `usb: class: cdc: Add ECM related macros` header additions, and the `usb: host: ch9: Add string descriptor related request helper function` ch9/desc helpers. I plan to pull these in (with credit) and replace the inline string parsing in this PR. That brings the two PRs structurally closer without merging the feature scope.

4. On the overall direction: my preference remains landing a minimal core first (this PR), with #100289 rebasing the additional features on top. This keeps each PR focused and easier to review but I would like @jfischer-no's call on which path is preferred before I proceed.

@jfischer-no could you weigh in on the preferred merge strategy for ECM host support?

tmon-nordic commented on subsys/usb/host/class/Kconfig.cdc_ecm_host:
```diff
@@ -0,0 +1,25 @@
+config USBH_CDC_ECM_INSTANCES_COUNT
+	int "Number of CDC ECM class instances"
+	default 1
+	range 1 4
```
Where does the upper limit of 4 comes from?

santhosh-c-c replied:
The number 4 is an arbitrary choice not tied to any spec.

tmon-nordic replied:
What about changing it to e.g. `range 1 $(UINT8_MAX)`?

santhosh-c-c replied:
Will change it to `range 1 $(UINT8_MAX)`

josuah commented on subsys/usb/common/ecm.c:
This is put in common but does not modify the device side class to use it,
is it intended for a future PR?

If merging it in this current form is planned, maybe this can stay in `usbh_cdc_ecm.c` and only move to `common` when/if some merging between device and host-side implementation is done?

santhosh-c-c replied:
The shared header was speculative, wrapped the contents directly into `usbh_cdc_ecm.c`

josuah commented on subsys/usb/host/class/usbh_cdc_ecm.c:
```diff
@@ -0,0 +1,753 @@
+		}
+	}
+
+	return ret;
```
Here, the `ret` value is not preserved from the body of the function in case `priv->state == CDC_ECM_CONFIGURED` is true. The body of the function never modifies it from 0.

It might be possible to:

- Turn `ret = 0;` into `ret;`
- `return 0;` at the end of the function

josuah commented on subsys/usb/host/class/usbh_cdc_ecm.c:
```diff
@@ -0,0 +1,753 @@
+	}
+
+	ret = cdc_ecm_fetch_mac(priv);
+	if (ret) {
```
Minor style element.

You might be asked during further review to preserve all `if()` expression as boolean. (point 85 [here](https://docs.zephyrproject.org/latest/contribute/coding_guidelines/index.html) FYI)

Typically, `if (ret != 0)`.

josuah commented on subsys/usb/host/class/usbh_cdc_ecm.c:
```diff
@@ -0,0 +1,753 @@
+	struct net_stats_eth stats;
+};
+
+static int cdc_ecm_start_rx(struct usbh_cdc_ecm_data *priv);
+static int cdc_ecm_start_intr(struct usbh_cdc_ecm_data *priv);
```
There does not seem to be circular dependencies, do you think it would be possible to reorder the functions so that this is not needed?

santhosh-c-c replied:
There's a cycle: `cdc_ecm_start_{rx,intr}` call `submit_xfer` with the `cdc_ecm_{rx,intr}_in_cb` callbacks, and the callbacks resubmit by calling `cdc_ecm_start_{rx,intr}`.

josuah replied:
I see now, thank you

josuah commented on subsys/usb/host/class/usbh_cdc_ecm.c:
```diff
@@ -0,0 +1,753 @@
+	while ((const void *)desc < desc_end && desc != NULL) {
+		switch (desc->bDescriptorType) {
+		case USB_DESC_INTERFACE:
+			if_desc = (const struct usb_if_descriptor *)desc;
```
Maybe it was meant to have the same exit condition as below in case the next USB function does not have an IAD?

```c
		case USB_DESC_INTERFACE_ASSOC:
			if (in_ecm_function) {
				goto done;
			}
```

josuah reviewed (commented):
Thank you for clarifying.

Would you be interested in helping #100289 rebase on top of this? For instance by indicating the parts that would need to be changed?

santhosh-c-c commented:
> Would you be interested in helping #100289 rebase on top of this? For instance by indicating the parts that would need to be changed?

Happy to help @josuah Here is how I see the migration mapping. Splitting into phases by dependency.

**Phase A: Independent PRs (both drivers benefit):**
* **ECM Macros:** Extract the `usb_cdc.h` additions from #100289 and land as a standalone PR.
* **CH9 Helpers:** Extract the string descriptor helpers (`usbh_req_desc_str`, `usbh_desc_str_utf16le_to_ascii`, etc.) from #100289. Once merged, this PR (#99097) will rebase to use them in `cdc_ecm_fetch_mac`, replacing the manual UTF-16LE parsing.

**Phase B: Merge Minimal ECM Core (#99097):**
* Lands with probe, single-pass descriptor parsing, MAC fetch, intr-in carrier tracking, bulk in/out, and initial `SET_ETHERNET_PACKET_FILTER`.

**Phase C: Incremental Feature PRs (ported from #100289):**
To keep reviews manageable, #100289’s driver commit becomes follow-up PRs on top of #99097:
1.  Switch to memslab-pooled `xfer cb_priv` + `queued_xfers` list.
2.  Dynamic packet filter, Multicast filter, Hardware statistics.
3.  `CONNECTION_SPEED_CHANGE` notification + dynamic link speed.
4.  Concurrent TX queue (`sema` + `DATA_TX_CONCURRENT_NUM`).
5.  Disconnect xfer cleanup (drain `queued_xfers`).

**Reconciliation Details:**
* **Descriptor Parser:** #100289 strictly requires `HEADER_FUNC_DESC`. This PR is permissive (handles real-world devices missing the HEADER). I lean toward keeping it permissive, but open to discussion.
* **Iface Bring-up:** #100289 correctly defers `usbh_device_interface_set` until `eth_iface` start (admin-up). I currently do it at probe, but plan to absorb NXP's approach here before the rebase.
* **TX Path:** Addressed in Phase C (4) migrating from my mutex single-xfer to NXP's sema concurrent model for better throughput.

@jfischer-no @MarkWangChinese Let me know your thoughts.

josuah commented:
> 1. Switch to memslab-pooled xfer cb_priv + queued_xfers list.

I have the impression that this can be avoided altogether.

The thing is: when removing a class from the stack, all associated transfers also need to be removed, and the way to do this is to call dequeue() on each, but there is no way to get the list of all the transfers that are queued to a particular class.

This could be done at the stack level maybe, as then it is a lot less effort for each and every class (implemented with an array for UVC for instance).

https://github.com/zephyrproject-rtos/zephyr/blob/44853cf97498ecbdf941bb3648f04333c54b4bc2/subsys/usb/host/class/usbh_uvc.c#L116

Without this, transfers de-allocation upon class removal is not being handled?

MarkWangChinese commented:
> > 1. Switch to memslab-pooled xfer cb_priv + queued_xfers list.
> 
> I have the impression that this can be avoided altogether.
> 
> The thing is: when removing a class from the stack, all associated transfers also need to be removed, and the way to do this is to call dequeue() on each, but there is no way to get the list of all the transfers that are queued to a particular class.
> 
> This could be done at the stack level maybe, as then it is a lot less effort for each and every class (implemented with an array for UVC for instance).
> 
> https://github.com/zephyrproject-rtos/zephyr/blob/44853cf97498ecbdf941bb3648f04333c54b4bc2/subsys/usb/host/class/usbh_uvc.c#L116
> 
> Without this, transfers de-allocation upon class removal is not being handled?

Hi @josuah For disconnecting, the controller will callback all the xfers that are enqueued. For the case of video: stooping the stream https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/usb/host/class/usbh_uvc.c#L3022. I think it is the controller driver interface issue, we needs one controller interface api to release the resource of "dev + endpoint" if it is no longer used, then the related usb bandwidth can be released. If needed, we can talk here https://github.com/zephyrproject-rtos/zephyr/issues/101159. Thanks.

santhosh-c-c commented:
> Hi @josuah For disconnecting, the controller will callback all the xfers that are enqueued. For the case of video: stooping the stream https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/usb/host/class/usbh_uvc.c#L3022. I think it is the controller driver interface issue, we needs one controller interface api to release the resource of "dev + endpoint" if it is no longer used, then the related usb bandwidth can be released. If needed, we can talk here #101159. Thanks.

@josuah - Based on Mark's explanation, does the host stack actually guarantee that `removed()` runs before the callbacks `(EP_REQUEST)` for cancelled transfers? Since they use different message queues `(usbh_bus_msgq and usbh_msgq)` processed by separate threads, I don't see anything ensuring that specific order.

josuah commented:
I think this is not guaranteed: there is a thread for endpoint events and other events, they are split in different queues there:

https://github.com/zephyrproject-rtos/zephyr/blob/603fa4818e8faedec88364d2eb3f6c44eb52d763/subsys/usb/host/usbh_core.c#L33-L45

That suggests full concurrency, and events such as "remove this device and all the classes" might need to go through mechanism like "atomically remove all endpoints from all queues and ask drivers they processed/errored-out the ongoing buffer.

santhosh-c-c commented:
> I think this is not guaranteed: there is a thread for endpoint events and other events, they are split in different queues there:

Thanks for the clarification @josuah

santhosh-c-c commented:
@jfischer-no @tmon-nordic @josuah
Ping for review, Thanks!

josuah commented on subsys/usb/host/class/usbh_cdc_ecm.c:
I think there is a helper function that combines all three in one which you may be able to use.

josuah commented on subsys/usb/host/class/usbh_cdc_ecm.c:
https://github.com/zephyrproject-rtos/zephyr/pull/102941 seems to introduce string descriptor fetching, maybe this is useful to this PR as well?

We might expect that not only CDC ECM will need string fetching, so introducing it as a stack feature would makes sense?

josuah commented on subsys/usb/host/class/Kconfig.cdc_ecm_host:
In order to configure the number of instances, you might be able to automate it using this snippet (combining it with the `log_config` below)

```
module = USBH_CDC_ECM
module-str = "usbh cdc_ecm"
default-count = 1
source "subsys/logging/Kconfig.template.log_config"
source "subsys/usb/common/Kconfig.template.instances_count"
```

GeorgeCGV commented on subsys/usb/host/class/usbh_cdc_ecm.c:
nit: 
```suggestion
#define ECM_MAC_ADDR_LEN	NET_ETH_ADDR_LEN
```
from already included `zephyr/net/ethernet.h`

GeorgeCGV commented on subsys/usb/host/class/usbh_cdc_ecm.c:
Maybe worth making `hex_nibble` return an error (_i.e. `-EINVAL`_) upon an invalid value instead of `0` to protect against malicious devices?

GeorgeCGV commented on subsys/usb/host/class/usbh_cdc_ecm.c:
nit:
```suggestion
#define CDC_ECM_ETH_MAX_FRAME_SIZE	(NET_ETH_MTU + sizeof(struct net_eth_hdr))
```
from already included `zephyr/net/ethernet.h`.

Ideally, the host shall rely on `wMaxSegmentSize` (_from a device descriptor_).

santhosh-c-c committed:
usb: host: add usbh_xfer_alloc_with_buf() helper

Add the usbh_xfer_alloc_with_buf() inline helper to usbh_device.h,
wrapping uhc_xfer_alloc_with_buf() at the USB host device layer.

This mirrors the existing usbh_xfer_alloc() wrapper and allows
callers to allocate a transfer with an associated buffer size
directly through the usb_device handle.

Signed-off-by: Santhosh Charles <santhosh@linumiz.com>

santhosh-c-c commented:
@josuah @GeorgeCGV - Addressed the review comments and rebased on main.

Changes:

- Added a small preparatory commit introducing usbh_xfer_alloc_with_buf() in usbh_device.h, mirroring usbh_xfer_alloc() on top of the existing uhc_xfer_alloc_with_buf(), so the class driver no longer allocates the transfer and buffer separately.
- MAC address retrieval now uses the string descriptor request helpers from #102941 (usbh_req_desc_str() + usbh_desc_str_to_ascii() + hex2bin()). That PR has not merged yet, so it is carried here as a cherry-picked commit, I will drop it once #102941 lands. Depends-on: #102941.
- Instantiation is now Kconfig-count based (CONFIG_USBH_CDC_ECM_INSTANCES_COUNT + LISTIFY) instead of a devicetree compatible.
- Remaining style/naming comments addressed.

Please take another look when you get a chance

GeorgeCGV commented on subsys/usb/host/class/usbh_cdc_ecm.c:
```diff
@@ -0,0 +1,788 @@
+	int ret;
+
+	total_len = net_pkt_get_len(pkt);
+	if (total_len == 0 || total_len > CDC_ECM_ETH_MAX_FRAME_SIZE) {
+		return -EMSGSIZE;
+	}
```
Any plans to parse `wMaxSegmentSize` from the descriptor and use it instead of `CDC_ECM_ETH_MAX_FRAME_SIZE`; or better `MIN(CDC_ECM_ETH_MAX_FRAME_SIZE, <wMaxSegmentSize>)`?

santhosh-c-c replied:
ACK, will update this!

GeorgeCGV replied:
Just discovered #100289. Marked it as resolved since I’m not sure which PR the maintainers will pick.
